cli/parse_test.ts

```diff
@@ -0,0 +1,725 @@
+    });
+    assertEquals(result, expected);
+  });
+  await t.step("optional argument after option with multiple values", () => {
+    const expected = {
+      options: { foo: ["bar", "baz"] },
+      arguments: {},
+    };
+    const result = parse(["--foo", "bar", "baz"], {
+      options: [
+        { name: "foo", value: { name: "VALUE", multiple: true } },
+      ],
+      arguments: [
+        { name: "VALUE", optional: true },
+      ],
+    });
+    assertEquals(result, expected);
+  });
+  await t.step("multiple arguments", () => {
+    const expected = {
+      options: {},
+      arguments: { VALUE: "foo", VALUE2: "bar" },
+    };
+    const result = parse(["foo", "bar"], {
+      arguments: [
+        { name: "VALUE" },
+        { name: "VALUE2" },
+      ],
+    });
+    assertEquals(result, expected);
+  });
+  await t.step("multiple different arguments", () => {
+    const expected = {
+      options: {},
+      arguments: { VALUE: ["foo", "bar"] },
+    };
+    const result = parse(["foo", "bar"], {
+      arguments: [
+        { name: "VALUE", multiple: true },
+      ],
+    });
+
+    assertEquals(result, expected);
+  });
+  await t.step("missing argument", () => {
+    assertThrows(() => {
+      parse([], {
+        arguments: [{ name: "VALUE" }],
+      });
+    });
+  });
+  await t.step("optional argument", () => {
+    const expected = {
+      options: {},
+      arguments: { VALUE: "foo" },
+    };
+    const result = parse(["foo"], {
+      arguments: [
+        { name: "VALUE", optional: true },
+      ],
+    });
+
+    assertEquals(result, expected);
+  });
+  await t.step("missing optional argument", () => {
+    const expected = {
+      options: {},
+      arguments: {},
+    };
+    const result = parse([], {
+      arguments: [
+        { name: "VALUE", optional: true },
+      ],
+    });
+    assertEquals(result, expected);
+  });
+  await t.step("optional multiple argument", () => {
+    const expected = {
+      options: {},
+      arguments: { VALUE: ["foo", "bar"] },
+    };
+    const result = parse(["foo", "bar"], {
+      arguments: [
+        { name: "VALUE", multiple: true, optional: true },
+      ],
+    });
+    assertEquals(result, expected);
+  });
+  await t.step("missing optional multiple argument", () => {
+    const expected = {
+      options: {},
+      arguments: {},
+    };
+    const result = parse([], {
+      arguments: [
+        { name: "VALUE", multiple: true, optional: true },
+      ],
+    });
+    assertEquals(result, expected);
+  });
+  await t.step("fn() is called", () => {
+    let called = false;
+    parse(["foo"], {
+      arguments: [
+        {
+          name: "run",
+          fn: () => {
+            called = true;
+          },
+        },
+      ],
+    });
+    assertEquals(called, true);
+  });
+});
+
+Deno.test("parse() handles commands", async (t) => {
+  await t.step("fn() is called", () => {
+    let called = false;
+    parse(["run", "--foo"], {
```

Comment:
I think parse with command should return which command was called.

This `parse` should return `{ command: ["run"], options: {...}, arguments: {...} }`.

Other examples:

- `[]` - no command was called.
- `["foo"]` - foo command was called.
- `["foo", "bar"]` - a nested command was called.

You would be able to discriminate on this property to get the correct types of options for a particular command.